corehq/apps/app_manager/models.py

```diff
@@ -3943,25 +3941,21 @@ def __attachment_cache_key(self, name):
 
     def __set_cached_attachment(self, name, content):
         cache.set(self.__attachment_cache_key(name), content, timeout=60 * 60 * 24)
+        self._LAZY_ATTACHMENTS_CACHE[name] = content
 
     def __get_cached_attachment(self, name):
-        return cache.get(self.__attachment_cache_key(name))
-
-    def __remove_cached_attachment(self, name):
-        cache.delete(self.__attachment_cache_key(name))
-
-    def __store_lazy_attachment(self, content, name=None, content_type=None,
-                                content_length=None):
-        info = {
-            'content': content,
-            'content_type': content_type,
-            'content_length': content_length,
-        }
-        self._LAZY_ATTACHMENTS[name] = info
-        return info
+        try:
+            # it has been fetched already during this request
+            content = self._LAZY_ATTACHMENTS_CACHE[name]
+        except KeyError:
+            content = cache.get(self.__attachment_cache_key(name))
+            if content is not None:
+                self._LAZY_ATTACHMENTS_CACHE[name] = content
+        return content
 
     def put_attachment(self, content, name=None, *args, **kw):
-        self.__remove_cached_attachment(name)
+        cache.delete(self.__attachment_cache_key(name))
+        self._LAZY_ATTACHMENTS_CACHE.pop(name, None)
         return super(LazyBlobDoc, self).put_attachment(content, name, *args, **kw)
```

Comment:
This has the same race condition that we're fixing in save. Maybe it would be better to remove this method and rename `lazy_put_attachment` to `put_attachment`? The caching done by this class is always going to have race conditions if items are removed from the cache before writing to the blob db.

Another option is to change this to update the cache after `super(LazyBlobDoc, self).put_attachment(...)` returns. However, that may not be safe in an `atomic_blobs` context.